### DISCUSSION TITLE
Something wrong about PETSc and libmesh(PETSc was not found)

### DISCUSSION THREAD

**Discussion by wzx-zzsdad** [idaholab/moose#19491] [Q&A Getting Started] [answered]
![image](https://user-images.githubusercontent.com/62682654/143732500-754e6edd-f6b3-44bf-ab70-08083696ab84.png)
When I use ./update_and_rebuild_libmesh.sh ,it shows me this question above. But my PETSc was already installed. I downloaded PETSc 3.13.2 and used this way to install PETSc below:
![image](https://user-images.githubusercontent.com/62682654/143733114-20bf9f1f-5862-4a04-a9e5-bd361bc89b58.png)
and the result of installing PETSc is below:
![image](https://user-images.githubusercontent.com/62682654/143732843-71c973c7-d645-42bb-9d24-3ca3447654cb.png)
my compiler and mpi information(I can use the same way to install moose successfully on other platform(openeuler) but on centos7.6 it can not be installed successfully):
![image](https://user-images.githubusercontent.com/62682654/143732855-f132c3b9-232d-417f-8239-3592f9757a14.png)
![image](https://user-images.githubusercontent.com/62682654/143733177-40b22674-8a73-4bc1-81b8-8d839b3babc6.png)
By the way,this information about compiling libmesh on openeuler platform is yes:
![image](https://user-images.githubusercontent.com/62682654/143733042-d8387b02-d111-4e34-b910-19624b52194a.png)

Please help me,thank you!

**Comment by milljm** [comment, marked as answer]
In the same directory where you ran `./update_and_rebuild_libmesh.sh`, there is a `libmesh_diagnostic.log`. Can you attach that to this thread? It gets created when you run ./update_and_rebuild_libmesh.sh. So if the diagnostic log is not present, just run the script again, let it fail, and then the log file should be present.

Also, in `/home/MOOSE/moose/libmesh/build`, there should be a config.log. Please attach that as well (the more information the better).

### Some thoughts

Flang could also be an issue if that is your selected fortran wrapper/compiler. To my knowledge, we have yet to make use of this new fortran compiler from the LLVM toolchain.

A simple test for this, would be to try and build libmesh without PETSc support (just to see if libMesh can be built using your compiler stack). Perhaps libMesh is using system compilers, while PETSc was instructed to use special compilers (--with-mpi-dir=/path/to/install/ompi). Being that is a screen shot, and not pasted text, I have to assume `/path/to/ompi` literally exists.

**Reply by wzx-zzsdad** [reply]
I rebuild libmesh in a new path, two logs:
[libmesh_diagnostic.docx](https://github.com/idaholab/moose/files/7622514/libmesh_diagnostic.docx)
[config.docx](https://github.com/idaholab/moose/files/7622515/config.docx)
I also tried  your thoughts.I build libmesh without PETSc but it will generate too many errors when 'make'.
![image](https://user-images.githubusercontent.com/62682654/143979807-7a603c5d-930d-4a0e-84bb-3742ebf0f0e5.png)
And I dont know where can I find complete information about this. 
Furthermore , if my system compilers is GCC9.1.0, I can successfully install libmesh by` ./updata_and_build_libmesh.sh` and moose even though my openmpi was installed by clang/flang. I tried specifying --with-cc=clang --with-cxx=clang++ --with-fc=flang in `configure_libmesh.sh` but it also cannot make successfully .Is there any other ways to install libmesh by specifying the compiler?

**Reply by milljm** [reply]
I don't have any Microsoft products installed... can you attache a regular .txt log?

[edit] never mind. I see it is only the suffix being called .docx. I can open these files just fine with a basic editor.

**Reply by milljm** [reply]
The error is with this:

```pre
configure:48980: result: <<< Found PETSc 3.13.2 installation in /home/MOOSE-clang/moose/petsc ... >>>
configure:48990: checking whether we can compile a trivial PETSc program
configure:49021: /path/to/install/ompi/bin/mpicxx -c  -I/home/MOOSE-clang/moose/petsc/include -I/home/MOOSE-clang/moose/petsc/arch-moose/include -I/home/MOOSE-clang/moose/petsc/arch-moose/include -I/home/MOOSE-clang/moose/petsc/include -I/home/MOOSE-clang/moose/petsc/arch-moose/include -I/path/to/install/ompi/include  conftest.cpp >&5
In file included from conftest.cpp:146:
In file included from /home/MOOSE-clang/moose/petsc/include/petsc.h:5:
In file included from /home/MOOSE-clang/moose/petsc/include/petscbag.h:4:
In file included from /home/MOOSE-clang/moose/petsc/include/petscsys.h:36:
In file included from /home/MOOSE-clang/moose/petsc/include/petscsystypes.h:218:
/usr/lib/gcc/aarch64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/complex:374:34: error: no matching member function for call to 'imag'
      complex<_Tp> __r(__x, -__y.imag());
                             ~~~~^~~~
/home/MOOSE-clang/moose/petsc/include/petsccxxcomplexfix.h:73:1: note: in instantiation of function template specialization 'std::operator-<double>' requested here
PETSC_CXX_COMPLEX_FIX(PetscInt)
^
/home/MOOSE-clang/moose/petsc/include/petsccxxcomplexfix.h:55:104: note: expanded from macro 'PETSC_CXX_COMPLEX_FIX'
static inline PetscComplex operator-(const Type& lhs, const PetscComplex& rhs) { return PetscReal(lhs) - const_cast<PetscComplex&>(rhs); } \
                                                                                                       ^
/usr/lib/gcc/aarch64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/complex:1224:7: note: candidate function not viable: 'this' argument has type 'const complex<double>', but method is not marked const
      imag() { return __imag__ _M_value; }
      ^
/usr/lib/gcc/aarch64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/complex:1245:7: note: candidate function not viable: requires single argument '__val', but no arguments were provided
      imag(double __val) { __imag__ _M_value = __val; }
      ^
/usr/lib/gcc/aarch64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/complex:1328:23: error: no matching member function for call to 'real'
          __real__ __t = __z.real();
                         ~~~~^~~~
/usr/lib/gcc/aarch64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/complex:435:11: note: in instantiation of function template specialization 'std::complex<double>::operator/=<double>' requested here
      __r /= __y;
          ^
/home/MOOSE-clang/moose/petsc/include/petsccxxcomplexfix.h:73:1: note: in instantiation of function template specialization 'std::operator/<double>' requested here
PETSC_CXX_COMPLEX_FIX(PetscInt)
^
/home/MOOSE-clang/moose/petsc/include/petsccxxcomplexfix.h:59:104: note: expanded from macro 'PETSC_CXX_COMPLEX_FIX'
static inline PetscComplex operator/(const Type& lhs, const PetscComplex& rhs) { return PetscReal(lhs) / const_cast<PetscComplex&>(rhs); } \
                                                                                                       ^
/usr/lib/gcc/aarch64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/complex:1220:7: note: candidate function not viable: 'this' argument has type 'const complex<double>', but method is not marked const
      real() { return __real__ _M_value; }
      ^
/usr/lib/gcc/aarch64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/complex:1242:7: note: candidate function not viable: requires single argument '__val', but no arguments were provided
      real(double __val) { __real__ _M_value = __val; }
      ^
/usr/lib/gcc/aarch64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/complex:1329:23: error: no matching member function for call to 'imag'
          __imag__ __t = __z.imag();
                         ~~~~^~~~
/usr/lib/gcc/aarch64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/complex:1224:7: note: candidate function not viable: 'this' argument has type 'const complex<double>', but method is not marked const
      imag() { return __imag__ _M_value; }
      ^
/usr/lib/gcc/aarch64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/complex:1245:7: note: candidate function not viable: requires single argument '__val', but no arguments were provided
      imag(double __val) { __imag__ _M_value = __val; }
      ^
3 errors generated.
configure:49021: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "libmesh"
| #define PACKAGE_TARNAME "libmesh"
| #define PACKAGE_VERSION "1.7.0-pre"
| #define PACKAGE_STRING "libmesh 1.7.0-pre"
| #define PACKAGE_BUGREPORT "https://github.com/libMesh/libmesh/discussions"
| #define PACKAGE_URL "http://libmesh.github.io"
| #define CONFIGURE_INFO "../configure  '--enable-silent-rules' '--enable-unique-id' '--disable-warnings' '--enable-glibcxx-debugging' '--with-thread-model=openmp' '--disable-maintainer-mode' '--enable-metaphysicl-required' '--with-cxx-std-min=2014' '--without-gdb-command' '--with-methods=opt oprof dbg' '--prefix=/home/MOOSE-clang/moose/scripts/../libmesh/installed' '--with-future-timpi-dir=/home/MOOSE-clang/moose/scripts/../libmesh/installed' 'INSTALL=/home/MOOSE-clang/moose/scripts/../libmesh/build-aux/install-sh -C' '--enable-petsc-hypre-required' 'PETSC_DIR=/home/MOOSE-clang/moose/petsc' 'PETSC_ARCH=arch-moose' 'METHODS=opt oprof dbg'"
```

I am guessing the compiler libMesh is trying to use is not the same as the one used to build PETSc. This coincides with what the libmesh_diagnostic.log also mentions:

```pre
Tue Nov 30 10:35:03 CST 2021
RUNNING AS ROOT

System Arch: CentOS Linux release 7.6.1810 (AltArch)

MOOSE Package Version: Custom Build

CPU Count: 128

Memory Free: 22509.125 MB

$CC not set                   <------------------ THIS SHOULD BE SET

MPICC:
which mpicc:
	/path/to/install/ompi/bin/mpicc
mpicc -show:
	clang -I/path/to/install/ompi//include/openmpi -I/path/to/install/ompi//include/openmpi/opal/mca/hwloc/hwloc201/hwloc/include -I/path/to/install/ompi//include/openmpi/opal/mca/event/libevent2022/libevent -I/path/to/install/ompi//include/openmpi/opal/mca/event/libevent2022/libevent/include -I/path/to/install/ompi//include -pthread -Wl,-rpath -Wl,/path/to/install/ompi//lib -Wl,--enable-new-dtags -L/path/to/install/ompi//lib -lmpi

COMPILER clang:
Bisheng Compiler 1.3.3.b023 clang version 10.0.1 (clang-e31092d4f8cd flang-3c92ea4b404f)
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/compiler/bisheng-compiler-1.3.3-aarch64-linux/bin
```

I believe libMesh will default to `gcc` if CC is not set... try exporting the following:

```pre
export CC=mpicc CXX=mpicxx FC=mpif90
```
and then run `./update_and_rebuild_libmesh.sh` again... If it fails, please attache the new logs again.

**Reply by milljm** [reply]
I see something else as well:

```pre
conftest.c:176:10: fatal error: '/home/MOOSE-clang/moose/petsc/arch-moose/include/petscversion.h' file not found
#include </home/MOOSE-clang/moose/petsc/arch-moose/include/petscversion.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

Does this location exist? Its possible you also need to export PETSC_ARCH= to whatever `arch-` dir is listed in `/home/MOOSE-clang/moose/petsc`. I have no way of guessing what that might be. If you are unsure, can you provide a directory listing of that location?

**Reply by wzx-zzsdad** [reply]
I tried `export CC=mpicc CXX=mpicxx FC=mpif90` and in the log $CC has set:
![image](https://user-images.githubusercontent.com/62682654/144163177-a266d6b9-4e73-47a8-b682-f2f31177f18c.png)
but also failed while installing libmesh.
[libmesh_diagnostic.log](https://github.com/idaholab/moose/files/7630415/libmesh_diagnostic.log)
[config.log](https://github.com/idaholab/moose/files/7630419/config.log)
And I have set :
```
export PETSC_DIR=/home/MOOSE-clang/moose/petsc
export PETSC_ARCH=arch-moose

```
But definitely there is no `petscversion.h` in my petsc path. I checked other machine which has installed moose successfully(with gcc-9.1.0 , the same configure parameters and PETSc version 3.13.2), there is no `petscversion.h`  in  `petsc/arch-moose/include` path either.

**Reply by milljm** [reply]
@fdkong When folks build petsc but not install it (not using --prefix), do users need to export PETSC_ARCH? I don't believe that directory exists, since there was nothing being installed.

Still, I am not sure that is the issue, since it looks like libMesh finds what it needs anyway:

```pre
checking /home/MOOSE-clang/moose/petsc/arch-moose/include/petscversion.h usability... no
checking /home/MOOSE-clang/moose/petsc/arch-moose/include/petscversion.h presence... no
checking for /home/MOOSE-clang/moose/petsc/arch-moose/include/petscversion.h... no
checking /home/MOOSE-clang/moose/petsc/include/petscversion.h usability... yes
checking /home/MOOSE-clang/moose/petsc/include/petscversion.h presence... yes
checking for /home/MOOSE-clang/moose/petsc/include/petscversion.h... yes
```
 
I am thinking libMesh configure is dying here:

```pre
configure:48980: result: <<< Found PETSc 3.13.2 installation in /home/MOOSE-clang/moose/petsc ... >>>
configure:48990: checking whether we can compile a trivial PETSc program
configure:49021: mpicxx -c  -I/home/MOOSE-clang/moose/petsc/include -I/home/MOOSE-clang/moose/petsc/arch-moose/include -I/home/MOOSE-clang/moose/petsc/arch-moose/include -I/home/MOOSE-clang/moose/petsc/include -I/home/MOOSE-clang/moose/petsc/arch-moose/include -I/path/to/install/ompi/include  conftest.cpp >&5
In file included from conftest.cpp:146:
In file included from /home/MOOSE-clang/moose/petsc/include/petsc.h:5:
In file included from /home/MOOSE-clang/moose/petsc/include/petscbag.h:4:
In file included from /home/MOOSE-clang/moose/petsc/include/petscsys.h:36:
In file included from /home/MOOSE-clang/moose/petsc/include/petscsystypes.h:218:
/usr/lib/gcc/aarch64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/complex:374:34: error: no matching member function for call to 'imag'
      complex<_Tp> __r(__x, -__y.imag());
                             ~~~~^~~~
```
But I am not sure why. PETSc seems to have built fine (and tested fine) in the original post.

**Reply by fdkong** [reply]
I am confused by several things here @wzx-zzsdad:

1) Originally, your PETSc was built as `PETSC_DIR=/home/MOOSE/moose/petsc PETSC_ARCH=arch-moose `

2) But then you set: 

```
export PETSC_DIR=/home/MOOSE-clang/moose/petsc
export PETSC_ARCH=arch-moose
```

Could you explain a bit more what the motivation here?


Libmesh error might be because you used (unintentionally) different compilers for libmesh and petsc.   

BTW, if you want to build PETSc, you really should use our in-house script: `./scripts/update_and_rebuild_petsc.sh.` If you want to point to a different MPI, you could do `./scripts/update_and_rebuild_petsc.sh --with-mpi-dir=balabala`

**Reply by wzx-zzsdad** [reply]
Firstly ,I git moose in a new directory so my PETSc path has changed. Secondly, I can use `./scripts/update_and_rebuild_petsc.sh` but I need to set `--with-openmp=0` in `configure_petsc.sh` if I use clang not gcc compiler. It will hang when `make check`  if I directlly run `./scripts/update_and_rebuild_petsc.sh`(default: `--with-openmp=1`) like this:
![image](https://user-images.githubusercontent.com/62682654/144339939-9286d8eb-f7f1-43f7-be9e-1873d39cafb4.png)
Even if I use` ./scripts/update_and_rebuild_petsc.sh ----with-mpi-dir=/path/to/install/ompi`, the installation of libmesh was also failed.
BTW,I can install moose successfully with gcc compiler, but there is no `petscversion.h` in arch-moose. 
![image](https://user-images.githubusercontent.com/62682654/144406159-fe7b3428-84a3-4f98-9630-93419953fc7d.png)

**Reply by milljm** [reply]
clang and clang++ should work, we use it all the time. But I am concerned about flang... That is, if when using the LLVM toolchain you are indeed using flang. Just so we all know, can you run the following:

```bash
mpif90 -show
```

**Reply by wzx-zzsdad** [reply]
![image](https://user-images.githubusercontent.com/62682654/144535025-f013a34d-2028-4c02-b16d-970c6c89c3fe.png)

**Reply by milljm** [reply]
Right, flang is probably going to be our issue. You are in uncharted waters using that compiler. I don't know how to fix this myself. Unless @fdkong has experience with it?
